Automattic/portkey-android
# Save frame service -  introduce StorySaveProcessor to handle state

This builds on top of PR #286, and will change the base branch once that one gets merged

Following a [comment](https://github.com/Automattic/portkey-android/pull/281#discussion_r398996970) by @jd-alexander (🙇) in #281, this PR includes the ability for the Service to handle new requests to save a Story while one Story saving process is already taking place.

For this we've introduced the concept of `StorySaveProcessor` and `StorySaveProcessorList` which hold the state keeping information for each story being saved, as opposed to the Service holding data for one unique Story each time, which would be otherwise overridden the next time it gets called, leading to corrupt information if the previous request was not yet already fully serviced.

Here's a video showing a Story with 6 videos and an image getting saved, and while the Service is still running we create a new Story containing one image, and get the Service to process / save it as well. 
https://cloudup.com/csNEDLyJgCq

#### To test
1. open the app
2. open the capture mode and add as many iamges and videos as you want (the more the better, given the idea is to try creating a new Story while the previous one is still getting saved).
3. remember to place at least one piece of text or emoji on top of each Story page, so these go through the saving process correctly
4. tap Next to trigger the Service for the first time.
5. now you'll be on the Main Activity. Tap on the FAB (+) button again to go into capture mode once more, for a new Story (you'll see the bottom strip only contains this new image you capture). Preferrably, capture an image as it's quick.
6. add some text to it
7. tap Next
8. drag from the top edge of the screen to see the foreground Service notification
9. wait for it to end, (in the meantime you can also observe the snackbars)
10. now go to the Photos app and check what has been saved. All of your new videos and images (both for the first and the second Story) should have been saved.
